### ISSUE TITLE
Multi-language support

### ISSUE THREAD

**Issue by turoksama** [lvgl/lvgl#562]
@kisvegabor @embeddedt
Hi,
  I think multi-language feature should be supported in the library. and i have a rough idea about this: 
1.add variables in ext structure who is associate with texts, eg. label object(base object of many widgets).
uint16_t ext->lang_index
char *ext->lang
char *ext->langs[6]
...

or

2:
c link variable, like: add custom fonts?

Is it possible to do that?

**Comment by kisvegabor** [comment]
I would be also interested in the feature. Why have you closed it?

**Comment by turoksama** [comment]
😄

**Comment by turoksama** [comment]
Hi,
When it is ready and I would like to raise a PR to do that.

**Comment by seyyah** [comment]
The simplest way is [1].

```c
int main() {
  lv_label_set_text(label, lang(ERROR_NETWORK_DISCONNECTED_YOUR_OPPONENT, esES) );
}
```

## :clipboard: Reference
1. https://gist.github.com/r-lyeh-archived/cbe7ebd681149af7b335b5f981651311

**Comment by seyyah** [comment]
Another way is [2]. I'm not sure gettext() can be used for littlevgl/embedded platform [2].

```c
#define _(STRING) gettext(STRING)

int main()
{
  /* Setting the i18n environment */
  setlocale (LC_ALL, "");
  bindtextdomain ("hello", "/usr/share/locale/");
  textdomain ("hello");

  /* Example of i18n usage */
  lv_label_set_text(label, _("Hello World\n"));

  return EXIT_SUCCESS;
}

// po/en/hello.po
#: hello.c:12
#, c-format
msgid "Hello World\n"
msgstr ""

// po/fr/hello.po
#: hello.c:12
#, c-format
msgid "Hello World\n"
msgstr "Bonjour le monde\n"
```

## :clipboard: Reference
1. http://www.labri.fr/perso/fleury/posts/programming/a-quick-gettext-tutorial.html
2. https://stackoverflow.com/questions/5300079/gettext-on-embedded-systems/5309936

**Comment by kisvegabor** [comment]
@seyyah 
> lv_label_set_text(label, lang(ERROR_NETWORK_DISCONNECTED_YOUR_OPPONENT, esES) );

I like it! It's clean and simple. 

@turoksama 
How/In which aspects can a built-in support better than this?

**Comment by turoksama** [comment]
@kisvegabor
Hi,
Yes, it's simple & cleared.
But i am looking for a way  to set language at run time. and i think it's automatically changed in system scale by setting a switch or something.
What do you think?

**Comment by seyyah** [comment]
```c
enum locale = enUS;

// a. TODO: change to locale at runtime

// b. set locale text
lv_label_set_text(label, lang(ERROR_NETWORK_DISCONNECTED_YOUR_OPPONENT, locale) );

// c. refresh the active window
// ??? lv_obj_invalidate(lv_scr_act());
```

**Comment by turoksama** [comment]
@seyyah
hmm...
What about button matrix?
It includes "\n" in the map of button matrix.

**Comment by seyyah** [comment]
Solution 1: locale string (for example `ERROR_NETWORK_DISCONNECTED_YOUR_OPPONENT`) contains `\n`.

Solution 2: you may use string library functionality [1]

```c
#include <string.h>

char str[N];
strcpy(str, lang(ERROR_NETWORK_DISCONNECTED_YOUR_OPPONENT, locale));
strcat(str, "\n");
strcpy(str, lang(ERROR_GENERIC_TEXT, locale));
...
```


## :clipboard: Reference
1. http://www.cplusplus.com/reference/cstring/strcat/

**Comment by turoksama** [comment]
nah,nah,nah...
the map text of button matrix looks like this:
if defined
char *button[]=
{
 "b1","b2","b3",""
}
the buttons matrix will palce in one row,
else if defined
```
char *button[]=
{
 "b1","\n",
 "b2","\n",
 "b3",""
}
```
they will place in one column.

of course you can store another pointer that point to the map.
and you can use that way to set language.

what i need is built in multi-language, all i have to do is tell the system where the language pointers are, and let the library do the rest.

i have no idea if i am worried too much.

And what do you suggest?

**Comment by turoksama** [comment]
my bad, didn't make myself cleared.
i prefer library behavior rather than user defined composition.
and i have achieved that goal already, but i dont like the feeling that things are messy.

**Comment by embeddedt** [comment]
@turoksama I think that in most cases the behavior suggested by @seyyah would be sufficient.

@kisvegabor Possibly we could add an alternative API for setting the button map on `lv_btnm` that makes multi-language support at runtime easier to implement?

**Comment by turoksama** [comment]
@embeddedt
Yes, it is.
If so, multiple changes should be made as far as i see.

**Comment by seyyah** [comment]
> So the question is how to extend the API of lv_btnm? What about `lv_btnm_set_map_str(btn, "a\tb\tc\n1\t2\t\n\t1\t2\t3")`. So `\t` (or other unused spcial character) could be the separator. What do you think?

I think that your suggestion and current implementation is based on text. It is simple/elegant but it can be insufficient for future requirements.

What about

```c
lv_btn_ext_t btnm_ext[];

lv_btnm_set_ext(btnm, btnm_ext);
```

**Comment by kisvegabor** [comment]
@seyyah I don't understand. Do you mean `lv_btnm_set_text()` Or what `lv_btn_ext_t` is?

**Comment by seyyah** [comment]
I recommend using `lv_btn_ext_t` for each sub-button of btnm. So the btnm_ext may be array, it's size is btnm's size

```c
lv_btn_ext_t btnm_ext[];
```

With this way, we can adjust each sub-button of btnm in detail as normal button.

Or alternatively may the btnm contain/be in array of btn?

**Comment by embeddedt** [comment]
@seyyah I believe the original intent of `lv_btnm` was to provide a way to use multiple buttons with little memory usage and low performance overhead.

By the time it is reworked to follow your suggestion I am afraid the memory requirements will be very similar to using multiple `lv_btn`s. Perhaps we should just add a note stating that `lv_btnm` does not support dynamic translation and users who require this feature should use `lv_btn` instead.

**Comment by kisvegabor** [comment]
@embeddedt Yes, the goal was to lower memory usage.

What about this:
```c
const char * btnm_map[10];
btnm_map[0] = lang(XYZ_1);
btnm_map[1] = lang(XYZ_2);
btnm_map[2] = lang(XYZ_3);
...
lv_btnm_set_map(btnm, btnm_map);
```

**Comment by embeddedt** [comment]
@kisvegabor For some reason I wasn't notified about your post. That looks like it could work but @seyyah or @turoksama should probably comment as well since they know more about it.

@nicolasvoron The final design has not yet been decided and after that someone will need to implement this. So at the moment I don't think anyone can answer your question.

**Comment by turoksama** [comment]
@embeddedt 
Hi,
 I am looking for a solution that integrated language system rather than manual ones.
If not so, why not just using:
```
const char *multi_lang[]=
{
  lang1,
  lang2,
  lang3,
  ...
}
```
and so on, there are many other such ways either.

**Comment by seyyah** [comment]
We're working on below solution (as draft),

```c
// Define lang keywords
#define LANG_STR1 0
#define LANG_STR2 1

// create lang strings
char lang_tr[N][M] = {"metin1", "metin2", "metin3", ...};
char lang_en[N][M] = {"string1", "string2", "string3", ...};

// Select lang
char (*lang_ptr)[M] = lang_tr;
char (*lang_ptr)[M] = lang_en;

// set lang string
lv_label_set_text(label, lang_tr[LANG_STR1]);
```

We have not be succeded yet.

## :clipboard: Reference
1. https://stackoverflow.com/questions/1052818/create-a-pointer-to-two-dimensional-array

**Comment by seyyah** [comment]
It works perfectly,

```c
#include <stdio.h>

#define N 2
#define M 3

int main(void) {
  char lang_TR[N][M] = {"aa", "bb"};
  char lang_EN[N][M] = {"11", "22"};

  char (*ptr)[M];

  printf("Language: TR\n");
  ptr = lang_TR;
  printf("matr: %s\n", lang_TR[0]);
  printf("matr: %s\n", lang_TR[1]);

  printf("ptr: %s\n", ptr[0]);
  printf("ptr: %s\n", ptr[1]);

  printf("Language: EN\n");
  ptr = lang_EN;
  printf("matr: %s\n", lang_EN[0]);
  printf("matr: %s\n", lang_EN[1]);

  printf("ptr: %s\n", ptr[0]);
  printf("ptr: %s\n", ptr[1]);
}
```

The output is

```bash
Language: TR
matr: aa
matr: bb
ptr: aa
ptr: bb
Language: EN
matr: 11
matr: 22
ptr: 11
ptr: 22
```

**Comment by seyyah** [comment]
Our littlevgl test code is

```c
#define N 2
#define M 3

char lang_TR[N][M] = {"aa", "bb"};
char lang_EN[N][M] = {"11", "22"};

char (*ptr)[M];

lv_obj_t *lbl_test = lv_label_create(lv_scr_act(), NULL);
lv_obj_set_pos(lbl_test, 10, 10);

ptr = lang_EN;
lv_label_set_text(lbl_test, ptr[0]);

ptr = lang_TR;
lv_label_set_text(lbl_test, ptr[0]);
```

**Comment by kisvegabor** [comment]
@seyyah I like this solution. I'd implement something similar too. It clear and easy to customize to work with any object type.

However, I think @turoksama wants it to work as a built-in method. According to your example: 
```c
char * dog_multi[2] = {"dog", "köpek"};
lv_label_set_text_multi(lbl_test, dog_multi);    /*Set all translations of "dog"*/

lv_lang_select(1);     /*Select Turkish and automatically refresh all labels and other elements*/
```

The question is it worth to add a built-in method for this or let to the user the make it himself.
If a built-in method would be added a new signal, .e.g. `LV_SIGNAL_CHANGE_LANGUAGE` should be added.

**Comment by BrianPugh** [comment]
Just to throw in my own 2 cents here.

This is how I implemented multilanguage support in Jolt external to lvgl:
https://github.com/joltwallet/jolt_wallet/tree/joltos/jolt_os/lang

Essentially every string for a given language in in it's one `.c` file (e.g. `english.c`). There's a enum type to index into the string list (so that we can refer to `STRING_HELLO_WORLD` rather than `3`). The num values are translated into a string via the `gettext` function which does a lookup according to the provided `enum` value and the currently selected language. Finally, each language has a list of fonts that it requires. Whenever a language is set, I store it to non-volatile storage and reset the program. 

I think it's a cleaner way of implementing multi-language because it:

- Keeps different languages in different files

- Allows compile-time inclusion/exclusion of language packs

- Allows compile-time inclusion/exclusion of font extensions.

Ultimately I think it would be nice to define a `lv_gettext(enum str_idx_t idx)` function in `lv_conf.h`. Labels (and similar text displaying objects) could then have a separate set function that takes in the `enum` instead of a string pointer. if `lv_set_language` is called, it could then send signals to all these objects to call `lv_gettext` again to get the new language's string.

**Comment by BrianPugh** [comment]
I agree your solution is easier/faster/more straight forward, but I also imagine that if someone is going through the trouble of adding localization to their code, they'll put in the extra effort to make it modular/maintainable. I'm just going to be typing up some ideas in no particular order here.

For 3, I think the text should always be obtained via a `lv_gettext(txt_id)` function, even if all it is doing is returning `lang_txt_array[txt_id]`. It allows for easier additional features in the future.

Just thinking aloud, i think each text object (labels, etc) should have a `union` in the `lv_ext` field like:
```
union {
    char *text;
    lv_text_id_t text_id;
}
uint8_t text_is_static:1;
```
where `text_is_static` is essentially a boolean on how to interpret the union. I see that the `lv_label` object has free bits for this. This will hide the 4 byte overhead that the `lv_text_id_t` would have consumed.

Calling `LV_SIGNAL_LANG_CHG` in `lv_label` will check if `text_is_static` is `true`/`false`. If the union contains a text_id, then it will re-call `lv_gettext(text_id)` which internally checks the currently set language.

For setting/checking if a language is supported, I currently have it set as a `uint64_t` bit field, where each bit represents a certain language. In this way, a program can easily check if a language is compiled into the project. This, of course, limits the number of languages to 64, but it could easily be a `uint8_t` array if more support is needed; I just used a `uint64_t` as a compromise between convenience and a realistic number of languages to support.

Each language pack needs the following:
1. An index into the language bitfield to let the system know which languages are available and which one is currently selected.
2. A list of `lv_font`'s that need to be added onto the core `lv_font` to support the language. Also, how long this list of additional `lv_font`s is.
3. The string array itself.

Finally, there needs be a function to trigger `LV_SIGNAL_LANG_CHG` on all objects to change the strings after a language change.

If there's any code you need me to contribute to aid in multi-language support, just let me know!

**Comment by kisvegabor** [comment]
> For 3, I think the text should always be obtained via a lv_gettext(txt_id) function, even if all it is doing is returning lang_txt_array[txt_id]. It allows for easier additional features in the future.

Sounds reasonable but I'd like to avoid adding functions directly into `lv_conf.h` for a high-level thing like this. 

What about `lv_lang_set(const char * (*lang_get_txt)(uint32_t txt_id)`? So to change language set a function pointer which can get the text of the language. It sounds even more flexible and modular because the languages can have their own `get_text` function. I don't see its clear beneifts right know but it seems the most future proof appoarch.

> Just thinking aloud, i think each text object (labels, etc) should have a union in the lv_ext field like...

The `text` field is still required because the `label` object assumes that it is dynamically allocated and sometimes it is modified (E.g. replace some character with dots).

> For setting/checking if a language is supported, I currently have it set as a uint64_t bit field, where each bit represents a certain language. In this way, a program can easily check if a language is compiled into the project. This, of course, limits the number of languages to 64, but it could easily be a uint8_t array if more support is needed; I just used a uint64_t as a compromise between convenience and a realistic number of languages to support.

I was thinking a little bit differently using the `lv_lang_set(function_pointer)` I mentioned befor. 

Add a typedef in lvgl like:
```c 
  typedef struct {
	  const char * name;		/*E.g. "English"*/
	  const char * (*get_text)(uint32_t id);
  }lv_lang_t;
``` 
You can build an array from `lv_lang_t` variables based on what is enabled and what isn't:
```c
  lv_lang_t languages[] = {
#if LANG_EN
		  {"English", en_get_text},
#endif
		  
#if LANG_ES
		  {"Spain", es_get_text},
#endif
		  {"", NULL},	/*Close*/
  };
```
This way it's easy to list the languages and get the `get_text` function.

> Each language pack needs the following:
> 1. An index into the language bitfield to let the system know which languages are available and which one is currently selected.
> 2. A list of lv_font's that need to be added onto the core lv_font to support the language. Also, how long this list of additional lv_fonts is.
> 3. The string array itself.

1. According to the above, I'm not sure it's required. But waiting for your answer! :)
2. As you can have a lot of styles related to labels/text it should be the user's responsibility to change/update the styles when the language is changed
3. I agree, the arrays are required :)

**Comment by BrianPugh** [comment]
So I went with the bitmask for "language support" rather than a string because:

1. It's quicker to compare (not by much, however)
2. The language name will probably be in english, which might not match the language's native name. So regardless, you will have to have a list of strings in your language pack of language names in each language (Spanish vs Espanol).
3. Because of 2, it's merely an identifier value, so a bitmask is the most compact solution.

I also don't think `lv_gettext` needs to be set by the user (even though it offers more power). I think it should be configured similar to how fonts are where you just declare languages in your lv_conf.h and provide the appropriate `c` files in the appropriate boilerplate format.

**Comment by turoksama** [comment]
I think we should leave an interface for lv_get_text, but a standard lv_get_text by default.
At this point of view, it is possible to let user decide what to do at run time.

**Comment by BrianPugh** [comment]
I agree

**Comment by kisvegabor** [comment]
> The language name will probably be in english, which might not match the language's native name. So regardless, you will have to have a list of strings in your language pack of language names in each language (Spanish vs Espanol).

Good point, I agree! In light of that my approach is really not so practical. But I'm still not sure the bitmap is required. With `ifdef`ing the languages and the elements of enum (like `jolt_lang_t`) should give the list of the languages indexable with the enum.  
Anyway, it's related to how you implement the language support in your code.  From `lvgl`'s point of view there should be:
- a `txt_id` in the related objects (e.g. `lv_label`)
- an `lv_lang_set` function to change the language which sends a signal to refresh all objects
- a function to set your `get_text` function. I still don't know a good name for it. `lv_lang_set_get_txt_func` seems odd because of the `_set_get_` part.

Do you agree with this summary?

**Comment by kisvegabor** [comment]
To see how it works in the practice I reworked the multi-lang support as discussed above. I tested it like this:
```c

typedef enum {
    LANG_EN,
    LANG_DE,
    LANG_HU,
}lang_t;

typedef enum {
    LANG_TXT_DOG,
    LANG_TXT_DAD,
}lang_txt_id_t;

const char * lang_en[] = {"Dog", "Dad"};
const char * lang_de[] = {"Hund", "Vater"};
const char * lang_hu[] = {"Kutya", "Apa"};

const void * lang_get_txt(uint16_t id)
{
    uint8_t lang = lv_lang_act();
    switch (lang) {
        case LANG_EN: return lang_en[id];
        case LANG_DE: return lang_de[id];
        case LANG_HU: return lang_hu[id];
        default: return NULL;
    }
}
...
    lv_lang_set_text_func(lang_get_txt);
    lv_obj_t * label = lv_label_create(lv_scr_act(), NULL);
    lv_label_set_text_lang(label, LANG_TXT_DOG);
    lv_lang_set(LANG_HU);
```

`lv_label_set_text_lang` doesn't look like a good name because it tells that the language of the label is set. Do you have a good idea for its name?

**Comment by seyyah** [comment]
Name: `lv_label_set_text_lang` -> `lv_label_set_text_lang_id`?

**Comment by BrianPugh** [comment]
> Good point, I agree! In light of that my approach is really not so practical. But I'm still not sure the bitmap is required. With `ifdef`ing the languages and the elements of enum (like `jolt_lang_t`) should give the list of the languages indexable with the enum.

So you'd have an array of enum's (lets say, typedef'd to be `uint8_t`) that hold the available languages? This is the exact same thing as the bitmap, just much less compact. To hold 64 languages, the bitmap takes 8-bytes. The array would take 64-bytes. I'm just bitshifting by the enum value. In your provided code example (which looks great by-the-way), how does the code know if german is available? Perhaps the german language was left out at compile-time due to space limitations. You could just remove it from the `lang_t` enum, but I'm not sure if this is a robust solution.

 Let's say the project running LVGL needs to interact with a program (or device) that wasn't available at compile-time. The application will inquire lvgl for the currently running language to determine localization features. LVGL could respond in a few ways:

1. Respond with the bitmask/array of available languages. It is assumed that the application knows the `enum` and languages will *only ever be appended to the enum*.
2. Respond with a string of the currently selected language (assumed to be the english name), but this is really just an inefficient version of (1) as now the application will need to hold (on average) about 500 extra bytes of strings just to compare languages (for 64-lang support).

However, to err on the side of simplicity, this whole "is language availalble" problem can be easily handled external to LVGL with a bitmask/array.

On a side note, this is an excellent example of where overriding lang_get_txt is useful. The application will have all of it's own language strings; yet, it might still be using core LVGL functions.

> Anyway, it's related to how you implement the language support in your code. From `lvgl`'s point of view there should be:
> 
> * a `txt_id` in the related objects (e.g. `lv_label`)
> * an `lv_lang_set` function to change the language which sends a signal to refresh all objects
> * a function to set your `get_text` function. I still don't know a good name for it. `lv_lang_set_get_txt_func` seems odd because of the `_set_get_` part.
> 
> Do you agree with this summary?

I completely agree with the summary.

> lv_label_set_text_lang doesn't look like a good name because it tells that the language of the label is set. Do you have a good idea for its name?

I think it should be `lv_label_set_txt_id`. I think the `lang` part is just confusing with actually setting the language (english/german/hungarian) for the object.

also `land_txt_id_t` should be `lang_txt_id_t`.

So, in summary, I'm happy with the new solution. I think a way of discovering "is language available" would be nice, but may end up just being implementation specific (which is probably the best, most powerful solution).

**Comment by puzrin** [comment]
IMHO, if you design built-in i18n support, it worth to care about plurals.

For small and medium-sized projects you may find useful notation from  https://github.com/nodeca/babelfish. It's battle tested and can be implemented via on-the-fly parser.

- For small projects, supported by "programmers", such approach is more comfortable than "gettext".
- For big projects, where i18n done by 3rd party hired translators, gettext is preferable because more familiar and better web tools availabiilty.

**Comment by kisvegabor** [comment]
@BrianPugh 
Thank you for the detailed description. This part seems really independent from lvgl so you can adjust it to your system and needs. Anyway, for a very simple case my view was this:
```c
typedef enum {
#if USE_LANG_EN
    LANG_EN,
#endif
#if USE_LANG_DE
    LANG_DE,
#endif
#if USE_LANG_HU
    LANG_HU,
#endif
}lang_t;

const void * lang_get_txt(uint16_t id)
{
    uint8_t lang = lv_lang_act();
    switch (lang) {
#if USE_LANG_EN
        case LANG_EN: return lang_en[id];
#endif
#if USE_LANG_DE
        case LANG_DE: return lang_de[id];
#endif
#if USE_LANG_HU
        case LANG_HU: return lang_hu[id];
#endif
        default: return NULL;
    }
}
```
For the new name of `lv_label_set_text_lang` simply `lv_label_set_text_id` seems good to me.

@puzrin 
I idea is good but it seems a little bit overkill for an embedded system in C.  Let's take this example line:
`I have #{count} ((nail|nails))`
1. Somehow the variable `count` need to be linked to the label. 
2. A format string might be useful: decimal point, leadning zeros. 

Anyway, parsing the texts is a possible extension in the future.

**Comment by puzrin** [comment]
>  `I have #{count} ((nail|nails))`
>
>  1. Somehow the variable count need to be linked to the label.

May be replace #{count} with {0}, and use params order. Should be ok in C.

Or add restriction - allow only one number  for one phrase, and use `#{}`. That will cover 99.9% of cases, and the rest can be done manually, as via `gettext()`.

>  2. A format string might be useful: decimal point, leadning zeros.

https://github.com/nodeca/babelfish/issues/18 possible markup extensions were documented, but i never needed to implemented such things in real world for JS.

**Comment by seyyah** [comment]
What do you think about this approach?

```c
const char * lang_en[] = {"Dog %d", "Dad %f"};
const char * lang_tr[] = {"%d Kopek", "%f Baba"};

const void * lang_get_txt(lv_obj_t * obj, uint16_t id)
{
    char buf[100];
    uint8_t lang = lv_lang_act();
    switch (lang) {
#if USE_LANG_EN
        case LANG_EN: 
            sprintf(buf, lang_en[id], lv_obj_get_variables(obj));
            return buf;
```

**Comment by kisvegabor** [comment]
@seyyah I like it! It means some extra work for the user compared to parsing texts automatically but provides great flexibility. In `lang_get_txt` you can build the text into `char txt_dad[32]` array too using `sprintf` and handle plural forms too.

If the "connected" variable changes it would be handy to have a function to refresh all labels with a given `txt_id`.

**Comment by embeddedt** [comment]
@kisvegabor @seyyah It looks like a good idea, as long as we don't require the user to use any functions from the `printf` family. Those usually take up quite a bit of flash.

**Comment by BrianPugh** [comment]
```
typedef enum {
#if USE_LANG_EN
    LANG_EN,
#endif
#if USE_LANG_DE
    LANG_DE,
#endif
#if USE_LANG_HU
    LANG_HU,
#endif
}lang_t;
```
Since this is just an enum, it's probably simpler to handle without all the `#if`. Also I like to start at 1 for these types of things because a 0 might mean something didn't get initialized properly.
```
typedef enum {
    LANG_EN=1,
    LANG_DE,
    LANG_HU,
}lang_t;
```
The `#if`'s in `lang_get_txt` handle the appropriate language logic to include at compile time.

**Comment by kisvegabor** [comment]
@embeddedt 
> @kisvegabor @seyyah It looks like a good idea, as long as we don't require the user to use any functions from the printf family. Those usually take up quite a bit of flash.

Form this point it's up to the user how to handle this. He can use `sprintf` or any custom "smaller" function

@BrianPugh 
I agree with both enums like this should start from 1 and the `#if`s are not required in the enum.

Regarding the function name I like `lv_label_set_text_id` the most. The comments before the function should clearly tell what the `_id` means.

**Comment by BrianPugh** [comment]
Here's an easy copy and paste language enum for you:
[lang.txt](https://github.com/littlevgl/lvgl/files/2796906/lang.txt)

**Comment by kisvegabor** [comment]
It seems to me everything is added as we discussed so I merged the `multi-lang` branch to `dev-5.3`.

Thank you for the collaboration! :clap: :clap: :clap:

**Comment by BrianPugh** [comment]
I'll start migrating my multi-lang code over to lvgl's in the next few days. Thanks for adding this to the code-base!

Some initial feedback:

For label, image, (and possibly others), you have the following pair of functions:

```
lv_label_set_text_id
```

```
lv_label_get_text_lang
```

It should be `lv_label_get_text_id`.

`lv_img_set_src_id` prototype is defined twice.

Throughout the code, language id is a `uint16_t`; a `uint8_t` should be more than sufficient, but I suppose a `uint16_t` maintains more flexibility and very little cost.

Also, what was on the conclusion on using formatted strings, like `"test number %d\n"`

**Comment by kisvegabor** [comment]
@puzrin 
I like that you are thinking about big things! :)
These really seem useful stuff for big projects. At first look, some of the ideas are related to external tools which can be developed independently from LittlevGL.

So please open a new issue and discuss it there!

**Comment by Next-connectmangrove** [comment]
I realized the solution



https://stackdev.space/questions/the-strcat-function-i-created-using-pointers-is-not-functioning-properly
